### PR TITLE
Draft: Introduce q-implant

### PR DESCRIPTION
**Pull request by Judgement9882**
This will introduce Quantization Parameter Importer(q-implant) and support int8 format in q-implant

ONE-DCO-1.0-Signed-off-by: Junyeong Kang kjy6135@naver.com

### CODE REVIEW

**Comment by Judgement9882** [comment]
for #11254, #11529
draft of q-implant : #11385

**Comment by seanshpark** [comment]
plz split PR for infra from compiler/q-implant codes

**Comment by seanshpark** [comment]
Q) #11385 is by @jinevening but why did @Judgement9882 post this PR ?

**Comment by Judgement9882** [comment]
> Q) #11385 is by @jinevening but why did @Judgement9882 post this PR ?

I'm a member of SSAFY's SSDC project. We are developing q-implant based on #11385. It hasn't been merged yet, so I thought I could post the PR together. I will put PR again after #11385 draft merged.

**Comment by Judgement9882** [comment]
I will reopen after #11385 merged.

**Comment by jinevening** [comment]
@seanshpark As talked offline, #11385 is a skeleton code, and this PR will complete `q-implant`.

@Judgement9882 Please proceed :) When you make non-draft PRs, please make separate PRs for the changes in `infra` and the changes in `compiler/q-implant`, as @seanshpark mentioned.

**Comment by Judgement9882** [comment]
@jinevening OK. I will pay attention to the points you mentioned.

**Comment by Judgement9882** [comment]
@nnfw-bot test nncc-release

**Comment by seanshpark** [comment]
I do not see any unit tests fro this project. 
Are there any plans?

**Comment by Judgement9882** [comment]
> I do not see any unit tests fro this project. Are there any plans?

We are currently designing unit tests for q-implant. 
As soon as it's done, we'll write the method on the "Issues" tab.

File: compiler/q-implant/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    "dtype": "uint8",
+    "scale": "0.npy",
+    "zerop": "1.npy",
+    "quantized_dimension": 0
+  },
+  ...
+}
+```
+
+`0.npy` and `1.npy` must exist in the same directory with `qparam.json`.
```
Comment (seanshpark) [inline compiler/q-implant/README.md:25]: moved from https://github.com/Samsung/ONE/pull/11579#pullrequestreview-1656684905

from this explanation, there will be two `.npy` files for one operator(node) from the model. for real models, there could be 10s or 100s for nodes in a model.
which would produces lots of files in the process of `q-implant`.
I don't think this is a good approach for users.

How about using `h5` format and produce single file for a model?

File: compiler/q-implant/tests/gen_test_data.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
```
Comment (seanshpark) [inline compiler/q-implant/tests/gen_test_data.py:1]: is this file still valid?

**Comment by seanshpark** [comment]
> As soon as it's done, we'll write the method on the "Issues" tab.

#11579 seems integration test. 

from https://github.com/Samsung/ONE/pull/11579#issuecomment-1746018928 , please add some unit function tests.

File: compiler/q-implant/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# q-implant
+
+_q-implant_ writes quantization parameters and weights (given as .json and .npy files) to a circle model.
```
Comment (seanshpark) [inline compiler/q-implant/README.md:3]: (moving discussion in #11579 to here)

plz explain more about "writes weights (.npy)"

CC @jinevening

Comment (seanshpark) [reply compiler/q-implant/README.md:3]: (moving discussion in https://github.com/Samsung/ONE/pull/11579/files#r1338134299 to here)

Then, q-implant should provide the structure / specification of this json and npy file.

File: compiler/q-implant/tests/gen_test_data.py (hunk 2 above)
Comment (jinevening) [reply compiler/q-implant/tests/gen_test_data.py:1]: There is an issue in using h5 format.

tensors can have / in their names, but h5 does not allow / in the object name. So we have to replace it into another character, which makes hard to recover the original tensor name.

---
Copied from https://github.com/Samsung/ONE/pull/11579#discussion_r1345284188

File: compiler/q-implant/README.md (hunk 1 above)
Comment (jinevening) [reply compiler/q-implant/README.md:25]: There is an issue in using h5 format.

tensors can have / in their names, but h5 does not allow / in the object name. So we have to replace it into another character, which makes hard to recover the original tensor name.

---
Copied from https://github.com/Samsung/ONE/pull/11579#discussion_r1345284188

File: compiler/q-implant/README.md (hunk 3 above)
Comment (jinevening) [reply compiler/q-implant/README.md:3]: Willl be discussed in https://github.com/Samsung/ONE/pull/11638.

File: compiler/q-implant/README.md (hunk 1 above)
Comment (seanshpark) [reply compiler/q-implant/README.md:25]: FYI, I used | character to use in replacement in our internal CI testing. not perfect but currently it's working.

Comment (seanshpark) [reply compiler/q-implant/README.md:25]: marked above comment as `off-topic` as there was some misunderstanding (README showed as an example)

File: compiler/q-implant/tests/gen_test_data.py (hunk 2 above)
Comment (Judgement9882) [reply compiler/q-implant/tests/gen_test_data.py:1]: > is this file still valid?

This file is not currently in use. 
However, we are currently applying the test data generation logic used in this file to each operator in the q-implant-op-level-test. 

(example : https://github.com/Samsung/ONE/pull/11579/files#diff-ca5cd73ec6907ee13c19d243d655c556a0d399187a8239df7c3d7e422d2c07ee)

**Pull request closed without merging**